### PR TITLE
CI/setup

### PR DESCRIPTION
**Pull request by lllAlexanderlll**
* CI for tests
* CI for linting
* Ci for build and deploy documentation (Wiki + API) to https://modalities.github.io/modalities/

### CODE REVIEW

File: config_files/config_lorem_ipsum.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ modalities_setup:
 
 wandb:
   project_name: modalities
-  mode: ONLINE
+  mode: OFFLINE
```
Comment (le1nux) [inline config_files/config_lorem_ipsum.yaml:8]: Shouldn't the default be `ONLINE`? Otherwise, the user has to manually call the syncing script.

File: docs/source/configuration.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
+.. role:: python(code)
```
Comment (le1nux) [inline docs/source/configuration.rst:1]: These files have the same content as the READMEs right? Is it possible to generate the rst files from the markdown? Otherwise, documentation always has to be updated in two places.

File: docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,53 @@
+Welcome to Modalities's documentation!
```
Comment (le1nux) [inline docs/source/index.rst:1]: Shouldn't it be "Modalities' documentation"? I don't want to be overly critical, but this is on the first page in the first line ... ;-)

File: tests/dataloader/test_packed_dataset.py [hunk 4]
```diff
@@ -26,8 +25,7 @@ def test_packed_megatron_dataset_loading(dummy_packed_data_path, block_size, exp
 def test_packed_continuous_dataset_loading(dummy_packed_data_path, block_size, expected_length, expected_output):
     ds = PackedMemMapDatasetContinuous(dummy_packed_data_path, block_size, sample_key="input_ids")
     assert len(ds) == expected_length
-    dl = LLMDataLoader(dataloader_tag="unittest", dataset=ds)
-    retrieved_input_ids = [list(x["input_ids"]) for x in dl]
+    retrieved_input_ids = [list(batch["input_ids"]) for batch in ds]
```
Comment (le1nux) [inline tests/dataloader/test_packed_dataset.py:28]: This change does not make sense to me. Why does `PackedMemMapDatasetContinuous` provide a `batch` in the for loop?

**Review by le1nux — CHANGES_REQUESTED** [review]
Apart from my comments, this looks good to me! I mainly covered the code part. 
@luzian-hahn could you take a look at the GitHub workflows and docs generation, please?

File: tests/dataloader/test_packed_dataset.py (hunk 4 above)
Comment (luzian-hahn) [reply tests/dataloader/test_packed_dataset.py:28]: I guess you are right, it would be rather `packed_samples`. Is the other change clear (the removal of the data_loader?

File: .github/workflows/build_and_deploy_documentation.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        pip install .
+        pip install sphinx-rtd-theme sphinx-autodoc-typehints sphinx-click sphinx-automodapi texext
+    - name: "Parse into HTML"
+      run: |
+        sphinx-apidoc -o docs/source/api src/modalities
+        sphinx-build -M html docs/source/ docs/build/
+    - name: Deploy to GitHub Pages
+      uses: peaceiris/actions-gh-pages@v3
+      with:
+        publish_branch: gh-pages
```
Comment (luzian-hahn) [inline .github/workflows/build_and_deploy_documentation.yml:30]: Does this mean, that the content is published to this branch? 
If so, isn't there the chance for conflicts when storing the documention on  parallel branch to main?

File: .github/workflows/build_and_deploy_documentation.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main 
+      - ci/setup
+
+permissions:
+  contents: write
```
Comment (luzian-hahn) [inline .github/workflows/build_and_deploy_documentation.yml:11]: Why is this necessary? Because we are pushing content to the branch "gh-pages"? If I understood this setting correctly, it indicates the permission to create "git-objects"

File: .github/workflows/build_and_deploy_documentation.yml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+name: "Build Sphinx Documentation"
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main 
+      - ci/setup
```
Comment (luzian-hahn) [inline .github/workflows/build_and_deploy_documentation.yml:8]: before merging, this should get removed

File: .github/workflows/build_and_deploy_documentation.yml (hunk 5 above)
Comment (luzian-hahn) [reply .github/workflows/build_and_deploy_documentation.yml:30]: okay I found your comment about this, that the `gh-pages` branch "listens" to `main`. What do you mean by that?

File: config_files/config_lorem_ipsum.yaml (hunk 1 above)
Comment (luzian-hahn) [reply config_files/config_lorem_ipsum.yaml:8]: I am not sure what this setting does. However we keep on encountering issues with WandB if the tests (this config is only for the unittests) run on different devices, which is always based on the fact that not everybody uses WandB (especially the CI test setup). Therefore deactivating this in the tests makes much sense in my opinion. 
I deactivated them beforehand by using a different subscriber, but if this here is also an option, I do very much appreciate it.

**Comment by luzian-hahn** [comment]
I fixed the tests for a CPU-only environment and[ tested them once again here in this PR](https://github.com/Modalities/modalities/actions/runs/7567511616/job/20606816686?pr=9).
Unfortunately I was only able to do so, by deactivating GPU-only features within the lorem_ipsum test scenario. If we want our CI-unittests to verify a working setup with GPUs, we might want to add another runner with GPUs to have the tests executed on it.

File: config_files/config_lorem_ipsum.yaml (hunk 1 above)
Comment (le1nux) [reply config_files/config_lorem_ipsum.yaml:8]: This setting stores the WandB logs only locally and does not sync them automatically. The user has to call a WandB syncing routine to push the logs to the WandB server. 

If this config is solely used for testing, then we can leave it as is. At the same time, we should have another config, which logs to WandB automatically to avoid confusion for the users.

File: tests/dataloader/test_packed_dataset.py (hunk 4 above)
Comment (le1nux) [reply tests/dataloader/test_packed_dataset.py]: yes the other changes are clear to me. I assume the intention behind them was to have the testing more granular? 
`packed_samples` would indeed be a better naming.

File: docs/source/configuration.rst (hunk 2 above)
Comment (lllAlexanderlll) [reply docs/source/configuration.rst:1]: - discussed in the group: Have the Wiki for tutorials separate. The readme should mainly contain our USPs.

File: .github/workflows/build_and_deploy_documentation.yml (hunk 5 above)
Comment (lllAlexanderlll) [reply .github/workflows/build_and_deploy_documentation.yml:30]: The Github action workflow listens to main and pushes the HTML files to the gh-pages branch.

Comment (lllAlexanderlll) [reply .github/workflows/build_and_deploy_documentation.yml:30]: We store the raw rst files on main, but not the HTML files. These will be created by sphinx and pushed only to the gh-pages branch

File: .github/workflows/build_and_deploy_documentation.yml (hunk 6 above)
Comment (lllAlexanderlll) [reply .github/workflows/build_and_deploy_documentation.yml:11]: Yes, because we want to create/write files in gh-pages

File: docs/source/index.rst (hunk 3 above)
Comment (lllAlexanderlll) [reply docs/source/index.rst]: Changed :)

File: .github/workflows/build_and_deploy_documentation.yml (hunk 7 above)
Comment (lllAlexanderlll) [inline .github/workflows/build_and_deploy_documentation.yml:8]:
```suggestion
```

Comment (lllAlexanderlll) [reply .github/workflows/build_and_deploy_documentation.yml]: Done

**Comment by le1nux** [comment]
LGTM :) Thanks for addressing the remarks.
